### PR TITLE
Improve the accuracy of asin(x) and asinh(x) for complex x using modified Hull et al algorithm.

### LINKED ISSUE #8553: Asinh not has accurate for complex64 dtype
```python
import numpy as np
import jax.numpy as jnp
from jax import jit
import torch
import paddle
import warnings

# Suppress warnings
warnings.filterwarnings("ignore")
# NumPy
complex_number_numpy = np.array([-1.0 - 101999.0j], dtype=np.complex64)
result_numpy = np.arcsinh(complex_number_numpy)
print("-------------------------------------------------------------------------------------------------------------------------------")
print("NumPy Result:", result_numpy)

import tensorflow as tf

# Define the complex number
complex_number = tf.constant([-1.0 - 101999.0j], dtype=tf.complex64)

# Calculate the inverse hyperbolic sine (asinh)
result = tf.math.asinh(complex_number)
print("---------------------------------------------------------------------------------------------------------------------------------------")
# Print the result
print("Tensorflow Result:", result.numpy())
# JAX
complex_number_jax = jnp.array([-1.0 - 101999.0j], dtype=jnp.complex64)
asinh_fn = jit(jnp.arcsinh)
result_jax = asinh_fn(complex_number_jax)
print("---------------------------------------------------------------------------------------------------------------------------------------")
print("JAX Result:", result_jax)

# PyTorch
complex_number_torch = torch.tensor([-1.0 - 101999.0j], dtype=torch.complex64)
result_torch = torch.asinh(complex_number_torch)
print("---------------------------------------------------------------------------------------------------------------------------------------")
print("PyTorch Result:", result_torch.numpy())

```
# Output 
```text
-------------------------------------------------------------------------------------------------------------------------------
NumPy Result: [-12.225865-1.5707865j]
---------------------------------------------------------------------------------------------------------------------------------------
Tensorflow Result: [-12.225865-1.5707865j]
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1705497661.625412   30592 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
[2024-01-17 18:51:01,626] [ WARNING] xla_bridge.py:527 - No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
---------------------------------------------------------------------------------------------------------------------------------------
JAX Result: [-5.348053+0.j]
---------------------------------------------------------------------------------------------------------------------------------------
PyTorch Result: [-12.225865-1.5707865j]

```

I think it's happening due to overflow at Asinh when using complex dtype . [here the source](https://github.com/openxla/xla/blob/64c6deaf73a80b5df90bd1e3c3df9ef5211fe976/xla/client/lib/math.cc#L1268-L1270)

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

~Fixes https://github.com/openxla/xla/issues/8553~ - PR https://github.com/openxla/xla/pull/9802 disabled the fix.

Update: the fix to https://github.com/openxla/xla/issues/8553 will be available via https://github.com/openxla/stablehlo/pull/2357

### CODE REVIEW

**Review by hawkinsp — COMMENTED** [review]
Could you add a regression test, in JAX python at least?

If you want to test this in JAX, it helps to bump `_version` in `xla_client.py` in this PR in the XLA repository, because then you can test that version number in JAX tests (search for `xla_extension_version` in the JAX tree).

**Comment by kamaljeeti** [comment]
Hi @pearu , any update on this PR, can you provide the information requested ? Thanks.

**Comment by kamaljeeti** [comment]
Hi @pearu , gentle reminder! Thanks.

**Comment by pearu** [comment]
> Hi @pearu , gentle reminder! Thanks.

@kamaljeeti , thanks for the reminder! The reviewers request for the regression test has been addressed via https://github.com/google/jax/pull/19823 . This PR can be updated after https://github.com/google/jax/pull/20144 has landed.

**Comment by penpornk** [comment]
@pearu https://github.com/google/jax/pull/20144 has landed. Could you please help update the PR? :)

**Comment by pearu** [comment]
> @pearu [google/jax#20144](https://github.com/google/jax/pull/20144) has landed. Could you please help update the PR? :)

@penpornk, certainly :) Meanwhile, https://github.com/openxla/xla/pull/10503 introduced a testing framework for functions with complex inputs in the XLA side. We should use this framework in this PR for asinh as well.

**Comment by penpornk** [comment]
Thank you for the heads up! https://github.com/openxla/xla/pull/10503 is now merged. (Many thanks to @reedwm!)

**Comment by kamaljeeti** [comment]
Hi @ddunl , there are a couple of checks failed on the Cl, can you have a look at them once? Thanks.

**Comment by pearu** [comment]
While using asin from asinh for complex inputs is the right way to go (as implemented in this PR), asin itself on complex inputs has very poor accuracy. I am currently implementing complex asin based on https://dl.acm.org/doi/10.1145/275323.275324 and then update this PR accordingly. Meanwhile, I'll mark the PR as a draft.

**Comment by kamaljeeti** [comment]
Hi @hawkinsp, can you have a look into this? Thanks.

**Comment by pearu** [comment]
As it turns out, asin/asinh improvements in `xla/client/lib/math.cc` will not be available for JAX. Therefore, I am currently working on submitting the modified Hull algorithm to StableHLO so that asin/asinh fixes here will be eventually effective also for JAX.

So, this PR should be considered as an improvement to asin/asinh in xla/client math lib that has no effect to the asin/asinh state in JAX.

File: xla/client/lib/math.cc [hunk 1]
```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
...
+                         Add(log2, Add(Log(mx), Mul(half, log1p_xoy))),
+                         Select(And(Lt(y, safe_min), Lt(x, one)), Div(y, sq),
+                                Log1p(Add(am1, sq))));
+      auto signed_imag = Select(Lt(signed_y, zero), Mul(none, imag), imag);
+
+      return Complex(real, signed_imag);
+    }
+    // TODO: handle overflow in x*x
+    return ScalarLike(z, 2.0) *
+           Atan2(x, ScalarLike(z, 1.0) + Sqrt(ScalarLike(z, 1.0) - z * z));
```
Comment (jakevdp) [inline xla/client/lib/math.cc:1295]:
```suggestion
           Atan2(z, ScalarLike(z, 1.0) + Sqrt(ScalarLike(z, 1.0) - z * z));
```

File: xla/client/lib/math.cc [hunk 2]
```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
...
+      auto smxm1 = Sub(s, xm1);
+      auto y_is_finite = Not(IsPosInf(y));
+      auto max_xy = Select(Gt(x, y), x, y);
+
+      auto y1 = Select(
+          Gt(max_xy, safe_max), y,
+          Select(
+              Le(x, one), Sqrt(Mul(half_apx, Add(Div(yy, rpxp1), smxm1))),
+              Mul(y, Sqrt(Add(Div(half_apx, rpxp1), Div(half_apx, spxm1))))));
+      auto real = Atan2(signed_x, y1);
```
Comment (jakevdp) [inline xla/client/lib/math.cc:1273]: I'm having trouble understanding this in terms of the formula in the comment above – I would have expected the real part of the return value to be `arcsin(x / a)`. Is the reasoning here that `y1` is constructed so that `x ** 2 + y1 ** 2 == a ** 2`, and therefore `arcsin(x / a)` is equivalent to `arctan(x/y1)`? If so, it may warrant a comment.

File: xla/client/lib/math.cc [hunk 3]
```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
...
+
+      auto am1 = Select(
```
Comment (jakevdp) [inline xla/client/lib/math.cc:1275]: Is this `a-1` with greater precision? If so a comment would be useful.

File: xla/client/lib/math.cc [hunk 4]
```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
...
-         Atan2(x, ScalarLike(x, 1.0) + Sqrt(ScalarLike(x, 1.0) - x * x));
+  XlaBuilder* b = x.builder();
+  auto do_it = [&](XlaOp z) -> absl::StatusOr<XlaOp> {
+    TF_ASSIGN_OR_RETURN(auto shape, b->GetShape(z));
+    if (primitive_util::IsComplexType(shape.element_type())) {
+      // According to "Implementing the complex arcsine and arccosine
+      // functions using exception handling" by Hull et al
+      // [https://dl.acm.org/doi/10.1145/275323.275324], complex
+      // arcsin function can be defined as
+      //   arcsin(z) = arcsin(x/a) + sign(y; x) * I * log(a + sqrt(a*a-1))
```
Comment (jakevdp) [inline xla/client/lib/math.cc:1198]: I'm assuming this is for `z = x + I*y`, yes?

**Review by jakevdp — COMMENTED** [review]
Looks reasonable to me, but I did not compare the code to the reference in detail. Hopefully the tests have enough coverage here...

File: xla/client/lib/math.cc (hunk 1 above)
Comment (pearu) [reply xla/client/lib/math.cc:1295]: Thanks for catching the typo! This actually indicates that arcus sine on real inputs is not tested in XLA.

File: xla/client/lib/math.cc (hunk 2 above)
Comment (pearu) [reply xla/client/lib/math.cc:1273]: First, the motivation for formulating the approximations from different regions in terms of `arctan2` is to reduce the number of calls to math functions to a minimal.
In the Hull et al paper, the real part is evaluated using a number of approximations that either call `arcsin` or `arctan` or simply `x / y`. It turns out that in specific regions, expressions with `arcsin` or division can be expressed in terms of `arctan2` and in other regions the approximations already use `arctan2`. Also, some regions using `arctan2` can be extended to cover certain parts of regions using `arcsin` without loosing accuracy for the given floating point system (float64, float32, ...).

While the Hull et al paper provides the mathematical proof to the implemented algorithm, there exists conditions where Hull et al are too conservative when considering that we are dealing with finite-precision numbers. MPMath implements the original Hull et al algorithm with multi-precision arithmetic that here we are using as reference for validating the modified Hull et al algorithm.

I'll add more comments explaining what is going on.

File: xla/client/lib/math.cc (hunk 4 above)
Comment (pearu) [reply xla/client/lib/math.cc:1198]: yes. I'll extend the comment accordingly.

File: xla/client/lib/math.cc (hunk 1 above)
Comment (pearu) [reply xla/client/lib/math.cc:1295]: > This actually indicates that arcus sine on real inputs is not tested in XLA.

Actually it does not because `z` and `x` are essentially the same objects.

File: xla/client/lib/math.cc (hunk 4 above)
Comment (pearu) [reply xla/client/lib/math.cc]: Fixed, an updated asin algorithm description is in math_impl.h

File: xla/client/lib/math.cc (hunk 3 above)
Comment (pearu) [reply xla/client/lib/math.cc]: `a - 1` approximation is now described in math_imph.h

File: xla/client/lib/math.cc (hunk 2 above)
Comment (pearu) [reply xla/client/lib/math.cc]: See asin algorithm description in math_impl.h, in particular the notes the real part.

File: xla/client/lib/math.cc (hunk 1 above)
Comment (pearu) [reply xla/client/lib/math.cc]: Fixed.

**Comment by pearu** [comment]
https://github.com/openxla/stablehlo/pull/2357 implements the modified Hull et al algorithm for StableHLO.

The asin implementations for both XLA and StableHLO are now generated using https://github.com/pearu/functional_algorithms tool that includes the validation of the algorithm against mpmath with the correctness of ULP difference <= 3. While in this PR we are still using tolerance conditions for measuring the difference of asin results and its reference values, using ULP difference is better measure because it does not depend on the magnitude of compared values. https://github.com/openxla/stablehlo/pull/2357 uses ULP difference, doing the same in XLA is planned as a follow-up.

**Comment by penpornk** [comment]
@jakevdp Could you please help take another look?

**Comment by dimitar-asenov** [comment]
@hawkinsp @jakevdp Could you please have a look at this PR and approve if it looks good to you?

**Comment by golechwierowicz** [comment]
@hawkinsp @jakevdp gentle ping

**Comment by penpornk** [comment]
@jakevdp Could you please help take another look at this PR? @pearu has addressed your initial comments.

Also, @GleasonK, would it make sense to request reviews from people who have reviewed https://github.com/openxla/stablehlo/pull/2357 ?

**Review by jakevdp — APPROVED** [review]
Looks fine to me, although I'm not familiar with the test infrastructure used here so I can't comment on that.

**Review by hawkinsp — APPROVED** [review]

**Comment by loislo** [comment]
I see that the function "static XlaOp Hypot(XlaOp x, XlaOp y)".
Is it intended? if it is not, could you please remove it?

File: xla/client/lib/math.cc [hunk 5]
```diff
@@ -1177,10 +1178,40 @@ XlaOp Acos(XlaOp x) {
   });
 }
 
+// hypot(x, y) = max(|x|, |y|) * sqrt(1 + (min(|x|, |y|) / max(|x|, |y|))^2)
+static XlaOp Hypot(XlaOp x, XlaOp y) {
```
Comment (loislo) [inline xla/client/lib/math.cc:1182]: The function is not in use/

Comment (pearu) [reply xla/client/lib/math.cc]: Done. Thanks!

**Review by loislo — APPROVED** [review]
